### PR TITLE
[CDEC-451] Remove `Mockable`

### PR DESCRIPTION
**Pull request by mhuesch**
## Description

The Mockable typeclass is a layer of indirection which allows code to be run in different monads without requiring changes to the monadic functions.

Mockable is apparently only executed in IO, and thus provides no benefit and a fair amount of hassle due to indirection & confusion for developers.

Remove the Mockable typeclass and associated instances.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CDEC-451

## Type of change
- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [~] 🛠 New feature (non-breaking change which adds functionality)
- [~] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist
- [~] I have added tests to cover my changes.
- [x] All new and existing tests passed.
^ I can see a good argument for tests to ensure the functions of `Pos.Core.Conc` are equivalent to those of `Pos.Core.Mockable.Production`. At present I have not added such tests, though.

### CODE REVIEW

**Comment by mhuesch** [comment]
Since there are a lot of changes here, I'd recommend going through the commits separately. The first 2 delete dead code which used `Mockable`. The next removes imports of `Mockable` from source code files. The next deletes the `Mockable` modules, and removes the last imports of `Mockable` from a few test packages.

File: networking/src/Network/Transport/ConnectionBuffers.hs [hunk 1]
```diff
@@ -1,290 +0,0 @@
...
-
-data InternalError =
-    InconsistentConnectionState
-  | UnknownConnectionClosed
-  | UnknownConnectionLost
-  | UnknownConnectionReceived
-  | DuplicateConnection
-  deriving (Typeable, Show)
-
-instance Exception InternalError
```
Comment (mhuesch) [inline networking/src/Network/Transport/ConnectionBuffers.hs]: It would be great to have @avieth confirm deletion of this module. It isn't included in the respective cabal file and seems fairly bitrotten.

File: infra/src/Pos/Infra/Discovery/Model/Neighbors.hs [hunk 2]
```diff
@@ -1,59 +0,0 @@
...
-    -> m ()
-converseToNeighbors which  sendActions handler = do
-    void $ withPeersConcurrently which $ \nodes -> do
-        logDebug $ "converseToNeighbors: sending to nodes: " <> show nodes
-        return $ \node -> do
-            handleAll (logErr node) $ withConnectionTo sendActions node $ \_ -> handler node
-            logDebug $ "converseToNeighbors: DONE conversing to node " <> show node
-    logDebug "converseToNeighbors: sending to nodes done"
-    where
-    logErr node e = logWarning $ sformat ("Error in conversation to " % shown % ": " % shown) node e
```
Comment (mhuesch) [inline infra/src/Pos/Infra/Discovery/Model/Neighbors.hs]: It would be great to have @avieth confirm deletion of this module. It isn't included in the respective cabal file and seems fairly bitrotten.

File: core/src/Pos/Core/Conc.hs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+-- toMicroseconds :: TimeUnit t => t -> Integer
+-- then we cast to an Int. Hopefully it fits!
+delay :: (TimeUnit t, MonadIO m)
+         => t
+         -> m ()
+delay time = liftIO (Conc.threadDelay (fromIntegral (toMicroseconds time)))
+
+-- | This function is analogous to `System.Timeout.timeout`, it's
+-- based on `Race` and `Delay`.
+-- TODO mhueschen - make sure this is equivalent to the old method
```
[the remark this answers is not in the document]
Comment (erikd) [reply core/src/Pos/Core/Conc.hs]: `timeout` certainly seems right to me.

File: infra/src/Pos/Infra/Slotting/Util.hs [hunk 4]
```diff
@@ -113,6 +113,8 @@ type MonadOnNewSlot ctx m =
 -- | Run given action as soon as new slot starts, passing SlotId to
 -- it.  This function uses Mockable and assumes consistency between
 -- MonadSlots and Mockable implementations.
+-- TODO mhueschen ^ get feedback on what this comment means and what
+-- we should do about it.
```
[the remark this answers is not in the document]
Comment (erikd) [reply infra/src/Pos/Infra/Slotting/Util.hs]: I *think* the comment you refer to is only relevant for `Mockable` instances were the underlying monad is not `IO`.

File: tools/src/dbgen/Main.hs [hunk 5]
```diff
@@ -142,9 +141,11 @@ newWalletState recreate walletPath =
     -- to rebuild the DB, but rather append stuff into it.
     liftIO $ openState (not recreate) walletPath
 
-instance HasLoggerName IO where
-    askLoggerName = pure $ LoggerName "dbgen"
-    modifyLoggerName _ x = x
+-- TODO mhueschen get feedback about usage of this module and whether
+-- the instance from `Pos.Core.Conc` can eclipse this.
```
[the remark this answers is not in the document]
Comment (erikd) [reply tools/src/dbgen/Main.hs]: This is in the `tools` package which basically just contains tools for the IOHK devops people to do what they need to do. 

On top of which, the current logging implementation is being replaced soon.

Not sure if that helps.

**Comment by CodiePP** [comment]
Looking forward to see this get merged into 'develop'. Getting rid of 'Production' will make our life easier!
Thanks for that great job @mhuesch

**Comment by coot** [comment]
LGTM, I am happy to see the change too.  As long as @avieth is fine with removing the files, I am fine with it.

File: auxx/Main.hs [hunk 6]
```diff
@@ -43,12 +42,12 @@ loggerName = "auxx"
 
 -- 'NodeParams' obtained using 'CLI.getNodeParams' are not perfect for
 -- Auxx, so we need to adapt them slightly.
-correctNodeParams :: AuxxOptions -> NodeParams -> Production (NodeParams, Bool)
+correctNodeParams :: AuxxOptions -> NodeParams -> IO (NodeParams, Bool)
```
Comment (parsonsmatt) [inline auxx/Main.hs]: :heart:

File: core/examples/abuse/Main.hs [hunk 7]
```diff
@@ -1,167 +0,0 @@
-{-# LANGUAGE BangPatterns       #-}
```
Comment (parsonsmatt) [inline core/examples/abuse/Main.hs]: Why is this file deleted?

File: core/examples/abuse/README.md [hunk 8]
```diff
@@ -1,177 +0,0 @@
-# Backpressure and peer abuse example
```
Comment (parsonsmatt) [inline core/examples/abuse/README.md]: Same question, why is it deleted? Seems to not be tied to the changes in this PR.

File: core/examples/abuse/Main.hs (hunk 7 above)
Comment (mhuesch) [reply core/examples/abuse/Main.hs]: `abuse` used `Mockable` & `Production`, and in trying to remove `Mockable` from it, I realized `abuse` was very bitrotten and contained imports of modules deleted from `cardano-sl` months ago. It seemed to me that the cost of updating the code would be significant, and that deleting it might be the best call. @avieth agreed with this, and I believe he was the last one to touch `abuse`.

I put a comment in the [git commit](https://github.com/input-output-hk/cardano-sl/pull/3285/commits/9007b703e09a3f8b1076f9b232b29418bdfd4973) but should've also written an explanatory comment on the PR.

File: infra/src/Pos/Infra/Slotting/Util.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mhuesch) [reply infra/src/Pos/Infra/Slotting/Util.hs]: Thanks, I think you're right. I found that the ["simple/default implementation"](https://github.com/input-output-hk/cardano-sl/blob/mhuesch/CDEC-451-sq/infra/src/Pos/Infra/Slotting/Impl/Simple.hs#L114) uses `currentTime`, which (before this PR) used to be a `Mockable` function. So there could be alternative implementations which might not have the appropriate consistency that the comment mentions.

I'll update the comment.

File: tools/src/dbgen/Main.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mhuesch) [reply tools/src/dbgen/Main.hs]: Thanks. I'll ask in the devops channel if anyone has advice or opinions on this.

File: core/examples/abuse/README.md (hunk 8 above)
Comment (avieth) [reply core/examples/abuse/README.md]: I'm OK with removing this. It's quite rotten.

File: infra/src/Pos/Infra/Discovery/Model/Neighbors.hs (hunk 2 above)
Comment (avieth) [reply infra/src/Pos/Infra/Discovery/Model/Neighbors.hs]: Yet another rotten file. Ok to delete.

File: networking/src/Network/Transport/ConnectionBuffers.hs (hunk 1 above)
Comment (avieth) [reply networking/src/Network/Transport/ConnectionBuffers.hs]: Yeah this was just a sketch and shouldn't have been committed at all.

File: tools/src/dbgen/Main.hs [hunk 9]
```diff
@@ -142,10 +141,6 @@ newWalletState recreate walletPath =
     -- to rebuild the DB, but rather append stuff into it.
     liftIO $ openState (not recreate) walletPath
 
-instance HasLoggerName IO where
```
Comment (CodiePP) [inline tools/src/dbgen/Main.hs]: this can savely be removed, as with the new logging all 'HasLoggerName' will disappear.
a LoggerName was used as a context name in logging. We will have a new mechanism which will cover this.

Comment (mhuesch) [reply tools/src/dbgen/Main.hs]: Great, thanks @CodiePP.
